File: client/jav/src/main/java/com/example/Consumer.java
package com.example;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import org.apache.kafka.clients.consumer.*;
import org.apache.kafka.common.serialization.StringDeserializer;

import java.time.Duration;
import java.util.Collections;
import java.util.Properties;

public class Consumer {
    private static final String TOPIC = "my-topic";
    private static final String BOOTSTRAP_SERVERS = "10.0.0.207:9094";
    private static final String GROUP_ID = "my-consumer-group";

    public static void main(String[] args) {
        Properties props = new Properties();
        props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, BOOTSTRAP_SERVERS);
        props.put(ConsumerConfig.GROUP_ID_CONFIG, GROUP_ID);
        props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
        props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
        props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
        props.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "true");
        
        // SASL SCRAM-SHA-512
        props.put("security.protocol", "SASL_SSL");
        props.put("sasl.mechanism", "SCRAM-SHA-512");
        props.put("sasl.jaas.config",
            "org.apache.kafka.common.security.scram.ScramLoginModule required " +
            "username=\"kafka-admin\" " +
            "password=\"...pw...\";");
        
        // TLS
        props.put("ssl.truststore.location", "kafka.truststore.jks");
        props.put("ssl.truststore.password", "changeit");

        ObjectMapper objectMapper = new ObjectMapper();
        objectMapper.registerModule(new JavaTimeModule());

        try (KafkaConsumer<String, String> consumer = new KafkaConsumer<>(props)) {
            consumer.subscribe(Collections.singletonList(TOPIC));
            System.out.println("Waiting for messages...");

            while (true) {
                ConsumerRecords<String, String> records = consumer.poll(Duration.ofMillis(100));
                
                for (ConsumerRecord<String, String> record : records) {
                    try {
                        Message message = objectMapper.readValue(record.value(), Message.class);
                        System.out.printf("Partition: %d, Offset: %d%n", 
                            record.partition(), record.offset());
                        System.out.printf("Value: %s%n%n", message);
                    } catch (Exception e) {
                        System.err.println("Error parsing message: " + e.getMessage());
                    }
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: client/sca/src/main/scala/com/example/Consumer.scala
package com.example

import org.apache.kafka.clients.consumer._
import org.apache.kafka.common.serialization.StringDeserializer
import io.circe.parser._
import java.time.Duration
import java.util.Properties
import scala.jdk.CollectionConverters._

object Consumer extends App {
  val topic = "my-topic"
  val bootstrapServers = "10.0.0.207:9094"
  val groupId = "my-consumer-group"

  val props = new Properties()
  props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers)
  props.put(ConsumerConfig.GROUP_ID_CONFIG, groupId)
  props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, classOf[StringDeserializer].getName)
  props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, classOf[StringDeserializer].getName)
  props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest")
  props.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "true")
  
  // SASL SCRAM-SHA-512
  props.put("security.protocol", "SASL_SSL")
  props.put("sasl.mechanism", "SCRAM-SHA-512")
  props.put("sasl.jaas.config",
    """org.apache.kafka.common.security.scram.ScramLoginModule required
      |username="kafka-admin"
      |password="...pw...";""".stripMargin)
  
  // TLS
  props.put("ssl.truststore.location", "kafka.truststore.jks")
  props.put("ssl.truststore.password", "changeit")

  val consumer = new KafkaConsumer[String, String](props)
  consumer.subscribe(List(topic).asJava)
  
  println("Waiting for messages...")

  try {
    while (true) {
      val records = consumer.poll(Duration.ofMillis(100))
      
      records.asScala.foreach { record =>
        decode[Message](record.value()) match {
          case Right(message) =>
            println(s"Partition: ${record.partition()}, Offset: ${record.offset()}")
            println(s"Value: $message\n")
          case Left(error) =>
            System.err.println(s"Error parsing message: ${error.getMessage}")
        }
      }
    }
  } catch {
    case e: Exception => e.printStackTrace()
  } finally {
    consumer.close()
  }
}